Refs #33826 -- Fixes an issue on the RedisCache backend delete_many and set_many methods

ckopanos opened the pull request:
Ref:

- https://code.djangoproject.com/ticket/33826

Changes:

- RedisCache backend changed delete_many and set_many methods to check for keys or data not been empty before performing an actual del or mset on redis

felixxm commented on django/core/cache/backends/redis.py:
```diff
@@ -218,15 +218,17 @@ def set_many(self, data, timeout=DEFAULT_TIMEOUT, version=None):
             key = self.make_and_validate_key(key, version=version)
             safe_data[key] = value
-        self._cache.set_many(safe_data, self.get_backend_timeout(timeout))
+        if safe_data:
+            self._cache.set_many(safe_data, self.get_backend_timeout(timeout))
```
Is there a reason to do this with `safe_data` instead of passed `data`? (the same in `delete_many()`), e.g.
```python
def set_many(self, data, timeout=DEFAULT_TIMEOUT, version=None):
    if not data:
        return []
    ...
```

ckopanos replied:
Not really no.. Now that i think of it this will also address issues where data or keys (delete_many) don't have the expected type.. e.g. keys been None or data been None..

Is it ok to modify the pr as per you suggestion and adjust the 2 tests accordingly to cover for keys or data been None?

felixxm replied:
> Not really no.. Now that i think of it this will also address issues where data or keys (delete_many) don't have the expected type.. e.g. keys been None or data been None..

Why? `data`/`keys` are not falsey when they contains `None` :thinking:

ckopanos replied:
I mean introducing your suggestions to place the check at the top of the method will also resolve issues where e.g data or keys parameter passed by the user in the respective methods is None.

In that case the relevant code block e.g

for key, value in data.items ....

Will fail. This is why it's a better approach.


I'll proceed with the changes.

Thanks for the suggestion

felixxm replied:
Ahh sorry, I misunderstood. Thanks :+1:

felixxm commented on tests/cache/tests.py:
```diff
@@ -1834,6 +1834,18 @@ def test_redis_pool_options(self):
         self.assertEqual(pool.connection_kwargs["socket_timeout"], 0.1)
         self.assertIs(pool.connection_kwargs["retry_on_timeout"], True)
 
+    def test_delete_many_no_keys(self):
+        # Behaviour when no keys are passed should not raise
+        # redis.exceptions.ResponseError: wrong number of arguments
+        # for 'del' command
+        self.assertIsNone(cache.delete_many([]))
+
+    def test_set_many_empty_data(self):
+        # Behaviour when an empty dict is passed should not raise
+        # redis.exceptions.ResponseError: Command # 1 (MSET)
+        # of pipeline caused error: wrong number of arguments for 'mset' command
+        self.assertEqual(cache.set_many({}), [])
```
I'd move these tests to the `BaseCacheTests` and remove comments.

ckopanos replied:
Ok.

Will check also if those 2 tests been placed to the basetest will introduce issues to the other backends and fix accordingly if needed

felixxm replied:
I checked locally. It should pass for other backends.

ckopanos replied:
I made the changes locally. By moving the tests to the base class and also extending them to check for keys or data being None, 

e.g.

```
self.assertIsNone(cache.delete_many(None))
```

tests fail cause core/cache/backends/base.py

does not account for keys or data been None.

Tests either fail with

AttributeError: 'NoneType' object has no attribute 'items'

TypeError: 'NoneType' object is not iterable.

Are you interested in covering for such a case in the base class and therefore do you want me to extend the check for keys and data in BaseCache or is this something that is out of the scope of the framework design? (in the sense that the developer should account for those cases when calling delete_many or set_many)?

ckopanos commented:
@felixxm I moved checking for empty keys or data to the top of the methods and moved the tests to BaseCacheTest as well.

I also did a minor improvement to the delete_many method of RedisCache to use a list comprehension when creating safe_keys which will perform better if the amount of initial keys exceeds a certain threshold
